### ISSUE TITLE
[FEATURE] Simplifing names

### ISSUE THREAD

**Issue by IceTrooper** [unity-atoms/unity-atoms#129]
It's a suggestion from me after seeing that:
"AtomBaseVariable Base Event Reference Listener"

**Describe the solution you'd like**

Listeners
---
AtomBaseVariable Base Event Reference Listener -> Atom Listener
\<T\> Event Reference Listener -> \<T\> Listener or \<T\> Atom Listener
\<T\>Pair Event Reference Listener -> <T>Pair Listener or \<T\>Pair Atom Listener
\<T\>\<K\> Event Reference Listener -> \<T\>\<K\> Listener or \<T\>\<K\> Atom Listener

Instancers
---
\<T\> Variable Instancer -> \<T\> Instancer
...

Event Instancers
---
AtomBaseVariable Event Instancer -> Atom Event Instancer
\<T\> Event Instancer
event instancers got right name

Sync
---
I'm not sure here.
![obraz](https://user-images.githubusercontent.com/51023959/79702140-e0707280-82a2-11ea-89eb-fde5ecd4b563.png)
Why does Sync \<T\> Variable Instancer to Collection got both properties: Sync To Collection and Sync to List?

Sync \<T\> To List -> \<T\> Sync or \<T\> Atom Sync
Sync \<T\> Variable Instancer to Collection -> \<T\> Instancer Sync or \<T\> Instancer Atom Sync

Other
---

Variable Resetter -> Atom Resetter

---

Most names are really long now and don't fit in context menu.

**Comment by AdamRamberg** [comment]
Thanks for this issue! The names are there also to be specific and to play nice with the generator. However, I think you have a valid point. I will think about this and come back to you.

**Comment by Casey-Hofland** [comment]
Might we also address the namespaces?
UnityAtoms.BaseAtoms > Atoms.Base

I would personally vote for `Atoms.Variable<float>` and `Atoms.Base.FloatVariable`.

**Comment by mgmhunt** [comment]
Using AtomTags - suggestion... rename most of the API with HasAtomTag vs HasTag for example....  spent quite a while debugging a HasTag vs CompareTag...  since AtomTags are self-contained to component then the method name should help reflect this. 

Not a big fan of moving 3rd party methods 'closer' to the UnityEngine so that they get confusing in IntelliSense and trying to become '1st class citizens'.. unless they are actually extending the Engine as such.  So the comment above 'Variable' vs Atom Variable... I would vote AtomVariable as both name spaces will be in use at the same time and Variable is too generic and also has potential to confuse with another library.  Would apply to the suggestions at the start too (ie Atom Listener over just Listener).

While I'm here (not related sorry!) - inspector for Atom Variables, make a lot more space for Value, rather than SO slot and give the GO name when has value rather than 'type mismatch' :)

**Comment by soraphis** [comment]
> Using AtomTags - suggestion... rename most of the API with HasAtomTag vs HasTag for example.... 

I agree, it should be Has/FindBy/Remove/...`AtomTag` instead of just Tag.

>  So the comment above 'Variable' vs Atom Variable... I would vote AtomVariable..

I also agree with that. I'm not against simplifying names, but we should not drop "Atom", on the contrary I think it could be nice - in terms of intellisense - start all end-user relevant classes with "Atom" (or atleast 'A').

That way new users would - at least on the code side - see quickly features/classes they did not knew existed. 

> While I'm here (not related sorry!) - inspector for Atom Variables, make a lot more space for Value, rather than SO slot and give the GO name when has value rather than 'type mismatch' :)

since it's not related, maybe just create a new issue for that? maybe even add an image to show what exactly is on your mind :)

**Comment by IceTrooper** [comment]
I have a completely different opinion. Why add the word "Atom" to each type if we are in the namespace Atoms? After all, that's what namespaces in C# are for. Methods should have a simple and clear name. In my opinion, they should even be close to those standard methods in Unity. Why should a method have the word "Atom" in the name member, if it is used on a type from Atoms (Atoms.SomeAtomTag.HasAtomTag() - looks strange, to say the least)?

I'm not totally against naming types starting with the word "Atom", e.g. "Atoms.AtomVariable" or "Atoms.AtomReference", because even though we are repeating ourselves relative to the namespace, I understand that this can be a helpful convention for some people (just like adding "_" at the beginning of private variable names, which I personally don't do, but I get it). Note, however, that maybe the base type will be called "AtomVariable", then later inheritances do not start with the word "Atom", but, for example, "FloatVariable", "BoolVariable". It would be possible to change the names to "AtomFloatVariable" or "AtomBoolVariable" so that they continue to start with the word Atom, but I think it reads badly. For this reason, this convention does not stick to itself.

As for method names, I don't think we need to specifically emphasize the word "Atom" once again, and in this case, I don't see any advantage in doing so.

**Comment by mgmhunt** [comment]
My issues was mainly with extension method on GO (ie at namespace level):

        go.tag = "Test";

        go.CompareTag("Test"); // Which one is the Atom Tag component vs go.tag?
        go.HasTag("Test");

        go.Atom.CompareTag("Test");  //This doesn't work 
        go.HasAtomTag("Test"); //Easier to distinguish
        go.AtomHasTag("Test"); // doesn't read well...?
        go.CompareAtomTag("Test"); // maybe closer to Unity... I don't like Compare but better to be closer to similar func?

For objects then yea Atom.Variable would work not too hung up on that :)  But kinda prefer not to type namespaces all the time...but yea probably should ;) Mainly the above method naming because I had trouble finding a bug related to the above.

**Comment by soraphis** [comment]
> Why add the word "Atom" to each type if we are in the namespace Atoms?

because the namespace does not give you anything regarding intellisense.

> Why should a method have the word "Atom" in the name member, if it is used on a type from Atoms...

1. as mgmhunt wrote, it's in this case mainly about extension methods on GameObjects.
2. even if looking beyond extension methods, issue is that unity has also a lot of stuff just on MonoBehaviour or Component, so even if you have a Component of Type AtomTag, there is "CompareTag" and "HasTag" on it. both can be called with a string parameter.

In C# you don't really care for the namespace at all. In C++ you'd write things like Atoms::AtomVariable, but in C# your IDE will usually automatically add all necessary using statements anyway and then you'll never see the namespace. 

> Note, however, that maybe the base type will be called "AtomVariable", then later inheritances do not start with the word "Atom", but, for example, "FloatVariable", "BoolVariable".

This for me is the exact opposite of what I'd consider useful :D

The Basetype isn't used ANYWHERE in the library-users code, so there we don't need to add "Atom", because the namespace is already enough. 
But for all the other types I'd personally want to reduce confusion as much as possible, so it should be clear to users what is Atoms, what is package A, B or C what comes from another library, the developers custom codebase or native unity.

Even though there are namespaces, the library-user would never see the "Atoms." of "Atoms.Variable". The namespace is imho purely for scoping reasons and does not resolve any naming confusion that may arise from reading over a piece of code. 

We could (not advocating for it, just as an argument) drop it altogether since in the packages are already using asmdefs and we could put everything that is considered internal into it's own asmdef, which would not be referenced by the library-users.

**Comment by soraphis** [comment]
Not only class/type namings, but also folder structure could take a pass.

**Actions/Conditions/Functions** - folder skips the "Atoms..." prefix, all folders only contain a single file (in Core package).
Could be one folder named "Callables" containing all 3 of them

**AtomListWrapper/AtomListWrapper.cs** should be in Utils/DataStructures, since it is a utility class used within the library for editor stylings.

**StackTrace/\*** should be in Utils/DataStructures/\*

**Attributes/** should be Utils/Attributes/.

**Extensions/** should be Utils/Extensions/.

**Types/** should be Utils/Types/

**"EventInstancers" + "VariablesInstancers" but "EventReferences" + "References"**.
References should be "VariableReferences"

Observables/BaseObservable is never used - i think.



I guess, I'd be the only one who cares about the folder structure, but I open it quite often here in github, and things like "References" being "VariablesReferences" and all the clutter folders get really annoying over time.

(I also don't really like it that the "Core" package, is sorted below "BaseAtoms" package ("AtomsCore" and "EssentialAtomTypes" would've been better) , but I guess this is a lost battle :D)

**Comment by soraphis** [comment]
Closing v5 PRs as there is no active development on v5 anymore
